src/System.Threading/src/System/Threading/ReaderWriterLock.cs

```diff
@@ -0,0 +1,1339 @@
+
+                    // Check for last signaled waiting reader
+                    Debug.Assert((knownState & LockStates.ReaderSignaled) != 0);
+                    Debug.Assert((knownState & LockStates.ReadersMask) < LockStates.ReadersMask);
+                    if ((knownState & LockStates.WaitingReadersMask) == LockStates.WaitingReader)
+                    {
+                        // Reset the event and the reader signaled flag
+                        readerEvent.Reset();
+                        Interlocked.Add(ref _state, -LockStates.ReaderSignaled);
+                    }
+
+                    break;
+                } while (YieldProcessor());
+            }
+
+            // Success
+            Debug.Assert((_state & LockStates.Writer) == 0);
+            Debug.Assert((_state & LockStates.ReadersMask) != 0);
+            ++lockEntry._readerLevel;
+#if RWLOCK_STATISTICS
+            Interlocked.Increment(ref _readerEntryCount);
+#endif
+        }
+
+        public void AcquireReaderLock(TimeSpan timeout) => AcquireReaderLock(ToTimeoutMilliseconds(timeout));
+
+        public void AcquireWriterLock(int millisecondsTimeout)
+        {
+            if (millisecondsTimeout < -1)
+            {
+                throw GetInvalidTimeoutException(nameof(millisecondsTimeout));
+            }
+
+            int threadID = GetCurrentThreadID();
+
+            // Check for the fast path
+            if (Interlocked.CompareExchange(ref _state, LockStates.Writer, 0) == 0)
+            {
+                Debug.Assert((_state & LockStates.ReadersMask) == 0);
+            }
+            // Check if the thread already has writer lock
+            else if (_writerID == threadID)
+            {
+                if (_writerLevel == MaxAcquireCount)
+                {
+                    throw new OverflowException(SR.Overflow_UInt16);
+                }
+                ++_writerLevel;
+                return;
+            }
+            else
+            {
+                int spinCount = 0;
+                int currentState = _state;
+                do
+                {
+                    int knownState = currentState;
+
+                    // Writer need not wait if there are no readers and writer
+                    if (knownState == 0 || knownState == LockStates.CachingEvents)
+                    {
+                        // Can be a writer
+                        currentState = Interlocked.CompareExchange(ref _state, knownState + LockStates.Writer, knownState);
+                        if (currentState == knownState)
+                        {
+                            // Only writer
+                            break;
+                        }
+                        continue;
+                    }
+
+                    // Check for too many waiting writers
+                    if ((knownState & LockStates.WaitingWritersMask) == LockStates.WaitingWritersMask)
+                    {
+                        Helpers.Sleep(1000);
```

Comment:
This should be rare enough as well, there would need to be 64K waiting writers to meet this condition.